autowarefoundation/autoware.universe
# fix(ntp_monitor): move chronyc command execution to a timer

## Description
The diagnostic topic of `ntp_monitor` sometimes may not be sent periodically, every 1 second.
The root cause is that the execution of the `chronyc tracking` command gets stuck when the network connection is poor.

So the execution of the `chronyc` will be separated from the main thread using a separate thread along with a timer.
In addition, a feature will be added to monitor timeouts in `chronyc` and to send a WARNING notification when a timeout occurs.
The timeout duration is set to 5 seconds by default.

## Related links

[TIER IV INTERNAL LINK TO JIRA](https://tier4.atlassian.net/browse/AEAP-668)

## Tests performed

1. Run Autoware.
2. Run rqt_runtime_monitor.
    ```console
    ros2 run rqt_runtime_monitor rqt_runtime_monitor
    ```
3. Verify OK is reported during normal operation.
![image](https://github.com/autowarefoundation/autoware.universe/assets/57388357/6199e3b2-0caa-41c8-a0f6-b8037e29de43)

4. Verify Warning is reported when `chronyc tracking` gets stuck.
![image](https://github.com/autowarefoundation/autoware.universe/assets/57388357/d747dcf7-0575-439b-83db-44b674c8650c)

## Notes for reviewers

None

## Interface changes

None

## Effects on system behavior

None

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
